test/cpp/api/parallel.cpp

```diff
@@ -0,0 +1,179 @@
+#include <torch/nn/parallel/data_parallel.h>
+#include <torch/nn/pimpl.h>
+#include <torch/tensor.h>
+
+#include <iostream>
+#include <memory>
+#include <utility>
+#include <vector>
+
+using namespace torch::autograd;
+using namespace torch::nn;
+
+TEST_CASE("Parallel/DifferentiableScatter", "[cuda]") {
+  Scatter scatter(
+      {torch::Device(torch::kCUDA, 0), torch::Device(torch::kCUDA, 1)});
+
+  auto input = torch::ones(10, torch::requires_grad(true));
+  auto output = scatter.apply({input});
+
+  REQUIRE(output.size() == 2);
+  REQUIRE(output[0].size(0) == 5);
+  REQUIRE(output[1].size(0) == 5);
+
+  auto sum = output[0].to({torch::kCUDA, 1}) + output[1];
+  sum.backward();
+
+  REQUIRE(input.grad().defined());
+  REQUIRE(input.grad().device().is_cpu());
+  REQUIRE(input.grad().sum().toCInt() == 10);
+}
+
+TEST_CASE("Parallel/DifferentiableGather", "[cuda]") {
+  Gather gather(torch::Device(torch::kCUDA, 1));
+
+  auto a = torch::ones(5, torch::requires_grad(true).device({torch::kCUDA, 0}));
+  auto b = torch::ones(5, torch::requires_grad(true).device({torch::kCUDA, 1}));
+  auto output = gather.apply({a, b});
+
+  REQUIRE(output.size() == 1);
+  REQUIRE(output[0].size(0) == 10);
+  REQUIRE(output[0].device() == torch::Device(torch::kCUDA, 1));
+
+  output[0].backward();
+
+  REQUIRE(a.grad().defined());
+  REQUIRE(a.grad().device() == torch::Device(torch::kCUDA, 0));
+  REQUIRE(a.grad().sum().toCInt() == 5);
+
+  REQUIRE(b.grad().defined());
+  REQUIRE(b.grad().device() == torch::Device(torch::kCUDA, 1));
+  REQUIRE(b.grad().sum().toCInt() == 5);
+}
+
+TEST_CASE("Parallel/Replicate", "[cuda]") {
+  Linear linear(3, 4);
+  auto replicas = parallel::replicate(
+      linear, {torch::Device(torch::kCUDA, 0), torch::Device(torch::kCUDA, 1)});
+  REQUIRE(replicas.size() == 2);
+
+  auto original_parameters = linear->parameters();
+
+  auto replica1_parameters = replicas[0]->parameters();
+  for (auto& parameter : replica1_parameters) {
+    REQUIRE(parameter->device() == torch::Device(torch::kCUDA, 0));
+  }
+  replicas[0]->to(torch::kCPU);
+  REQUIRE(replica1_parameters.size() == original_parameters.size());
+  for (size_t i = 0; i < original_parameters.size(); ++i) {
+    REQUIRE(replica1_parameters[i]->allclose(*original_parameters[i]));
+    REQUIRE(
+        replica1_parameters[i]->data().data<float>() !=
+        original_parameters[i]->data().data<float>());
+  }
+
+  auto replica2_parameters = replicas[1]->parameters();
+  for (auto& parameter : replica2_parameters) {
+    REQUIRE(parameter->device() == torch::Device(torch::kCUDA, 1));
+  }
+  replicas[1]->to(torch::kCPU);
+  REQUIRE(replica2_parameters.size() == original_parameters.size());
+  for (size_t i = 0; i < original_parameters.size(); ++i) {
+    REQUIRE(replica2_parameters[i]->allclose(*original_parameters[i]));
+    REQUIRE(
+        replica2_parameters[i]->data().data<float>() !=
+        original_parameters[i]->data().data<float>());
+  }
+}
```

Comment:
These two blocks look like an unrolled loop to me 😄

Reply:
I personally prefer to do this kind of duplication in test code as long as its not unreasonable, it makes debugging easier (e.g. the line information about where the assertion hit tells you more than if it was in a loop)